### PR TITLE
backport PipedStreamBuffer to 2.5.x

### PR DESCRIPTION
**Pull request by bpholt**
This backports #2383 to 2.5.x, to address #2037 for those who can't upgrade to 3.x yet.

### CODE REVIEW

File: io/src/test/scala/fs2/io/IoSuite.scala [hunk 1]
```diff
@@ -147,6 +147,19 @@ class IoSuite extends Fs2Suite {
         }
       }
     }
+
+    test("PipedInput/OutputStream used to track threads, fs2 reimplementation works") {
```
Comment (bpholt) [inline io/src/test/scala/fs2/io/IoSuite.scala:151]: This was adapted for CE2; it was originally:

```scala
    test("PipedInput/OutputStream used to track threads, fs2 reimplementation works") {
      readOutputStream(1024) { os =>
        IO.blocking {
          val t = new Thread(() => os.write(123))
          t.start
          t.join
          Thread.sleep(100L)
        }
      }.compile.drain.map(_ => assert(true))
    }
```

Comment (vasilmkd) [reply io/src/test/scala/fs2/io/IoSuite.scala:151]: I'll take a look when I'm back in front of my computer a little later.

**Review by vasilmkd — APPROVED** [review]

**Comment by mpilquist** [comment]
Hm, CI build failed: https://github.com/typelevel/fs2/runs/2895343064?check_suite_focus=true#step:6:3616

**Comment by vasilmkd** [comment]
I'm keeping an eye on the CI.

**Comment by vasilmkd** [comment]
Yeah, the issue persists. I'm working on it.

**Pull request merged**